olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java

```diff
@@ -263,6 +279,362 @@ public int hashCode() {
+                    logger.fine(String.format("Unknown Property of key: %s with type: %s", propertyEntry.getKey(), prop.getClass()));
+                }
+            }
+        }
+
+
+
+        private static <T> boolean checkPresenceAllMatch(Map<String, T> aMap, Map<String, T> bMap, BiPredicate<T, T> matchPair) {
+            Set<String> allKeys = new HashSet<>(aMap.keySet());
+            allKeys.addAll(bMap.keySet());
+            return allKeys.stream().allMatch(k -> {
+                Optional<T> maybeA = Optional.ofNullable(aMap.get(k));
+                Optional<T> maybeB = Optional.ofNullable(bMap.get(k));
+                if (maybeA.isPresent() && maybeB.isPresent()) {
+                    return matchPair.test(maybeA.get(), maybeB.get());
+                } else {
+                    if (maybeA.isPresent()) {
+                        logger.fine(String.format("Missing property: a: %s, no value for b", k));
+                    } else {
+                        logger.fine(String.format("Missing property: b: %s, no value for a", k));
+                    }
+                    return false;
+                }
+            });
+        }
+
+        /**
+         * Note that this does not check the {@code name} field deliberately.
+         */
+        @Override
+        public int hashCode() {
+            return Objects.hash(className, simpleProperties, listProperties, listClassProperties, mapProperties);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (o instanceof StructuralConfigurationData) {
+                StructuralConfigurationData that = (StructuralConfigurationData) o;
+                if (this.className.equals(that.className)) {
+                    boolean simpleMatch = checkPresenceAllMatch(this.simpleProperties, that.simpleProperties,
+                            DerefedProperty::equals);
+
+                    if(!simpleMatch) {
+                        logger.fine("SimpleProperties don't match: as: " + this.simpleProperties + " bs: " +that.simpleProperties);
+                    }
+
+                    boolean mapMatch = checkPresenceAllMatch(this.mapProperties, that.mapProperties,
+                            (aMap, bMap) ->
+                                    aMap.keySet().equals(bMap.keySet()) &&
+                                            aMap.keySet().stream()
+                                                    .allMatch(k -> aMap.get(k).equals(bMap.get(k))));
+
+                    if(!mapMatch) {
+                        logger.fine("MapProperties don't match: as: " + this.mapProperties + " bs: " + that.mapProperties);
+                    }
+
+                    boolean listMatch = checkPresenceAllMatch(this.listProperties,that.listProperties, (as, bs) -> {
+                       boolean eq = Util.bagEquality(as, bs);
+                        if(!eq) {
+                            logger.fine("bags not equal:\na: " + as.toString() +  "\nb: " + bs.toString());
```

Comment:
I think this should be `ListProperties not equal`, or `ListProperties not equal using bag equality`. I see there's more logging further down, but I think being specific here and in the class one will make it a little easier to see.